### PR TITLE
[serve] Use environment variable for `root_url` from runtime env

### PR DESCRIPTION
**Pull request by nikitavemuri**
## Why are these changes needed?

This PR checks the runtime environment for the environment variable for the `root_url` rather than looking at the global environment variables. This still supports `RAY_SERVE_ROOT_URL` being specified globally because the global environment variables are copied into the runtime environment.

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/serve/config.py [hunk 1]
```diff
@@ -170,8 +176,9 @@ def location_backfill_no_server(cls, v, values):
     @validator("root_url", always=True)
     def fill_default_root_url(cls, v, values):
         if v == "":
-            if SERVE_ROOT_URL_ENV_KEY in os.environ:
-                return os.environ[SERVE_ROOT_URL_ENV_KEY]
+            serve_root_url_env = ray.get(get_serve_root_url_env.remote())
+            if serve_root_url_env:
+                return serve_root_url_env
```
Comment (edoakes) [inline python/ray/serve/config.py:181]: I'd prefer to just do this in the controller instead, we can just do it out-of-band from the `HTTPOptions` validation.

File: python/ray/serve/api.py [hunk 2]
```diff
@@ -122,7 +122,7 @@ def shutdown_serve_client():
 
     @property
     def root_url(self):
-        return self._http_config.root_url
+        return ray.get(self._controller.get_root_url.remote())
```
Comment (edoakes) [inline python/ray/serve/api.py:125]: can you cache this so we don't need to make an RPC every time we log it? :)

**Review by edoakes — APPROVED** [review]
Looks good, I think there's one case we're still not handling here though: if the user sets the `SERVE_ROOT_URL_ENV_KEY` env variable on the client side when connecting via Ray client, if it isn't also set on the server then it won't be picked up. I think this is probably ok but let's at least document it where appropriate?

**Review by simon-mo — APPROVED** [review]

**Comment by nikitavemuri** [comment]
> Looks good, I think there's one case we're still not handling here though: if the user sets the `SERVE_ROOT_URL_ENV_KEY` env variable on the client side when connecting via Ray client, if it isn't also set on the server then it won't be picked up. I think this is probably ok but let's at least document it where appropriate?

Yes, makes sense - I updated the documentation for `start`. I believe this is the current behavior without this PR also, right?

**Pull request merged**